tests/unit/test_solvers.py

```diff
@@ -0,0 +1,376 @@
+            run, and keeping all other arguments the same.
+        4. Check the two calls to 'refine' yield that same result.
+        """
+        initial_coresubset, solver, _ = refine_problem
+        coresubset, state = solver.refine(initial_coresubset)
+        if use_cached_state:
+            coresubset_cached_state, recycled_state = solver.refine(
+                initial_coresubset, state
+            )
+            assert eqx.tree_equal(recycled_state, state)
+            assert eqx.tree_equal(coresubset_cached_state, coresubset)
+        self.check_solution_invariants(coresubset, refine_problem)
+
+
+class ExplicitSizeSolverTest(SolverTest):
+    """Test cases for solvers that have an explicitly specified 'coreset_size'."""
+
+    SIZE_MSG = "'coreset_size' must be a positive integer"
+    OVERSIZE_MSG = "'coreset_size' must be less than 'len(dataset)'"
+
+    @override
+    def check_solution_invariants(
+        self, coreset: Coreset, problem: Union[_RefineProblem, _ReduceProblem]
+    ) -> None:
+        super().check_solution_invariants(coreset, problem)
+        solver = problem.solver
+        if isinstance(solver, ExplicitSizeSolver):
+            assert len(coreset) == solver.coreset_size
+
+    @pytest.mark.parametrize(
+        "coreset_size,context",
+        [
+            (1, does_not_raise()),
+            (1.2, does_not_raise()),
+            (0, pytest.raises(ValueError, match=SIZE_MSG)),
+            (-1, pytest.raises(ValueError, match=SIZE_MSG)),
+            ("not_cast-able_to_int", pytest.raises(ValueError)),
+        ],
+    )
+    def test_check_init(
+        self,
+        solver_factory: Union[type[Solver], jtu.Partial],
+        coreset_size: Union[int, float, str],
+        context: AbstractContextManager,
+    ) -> None:
+        """
+        Ensure '__check_init__' prevents initialisations with infeasible 'coreset_size'.
+
+        A 'coreset_size' is infeasible if it can't be cast to a positive integer.
+        """
+        modified_solver_factory = eqx.tree_at(
+            lambda x: x.keywords["coreset_size"], solver_factory, coreset_size
+        )
+        with context:
+            modified_solver_factory()
+
+    def test_reduce_oversized_coreset_size(
+        self, reduce_problem: _ReduceProblem
+    ) -> None:
+        """
+        Check error is raised if 'coreset_size' is too large.
+
+        This can not be handled in '__check_init__' as the 'coreset_size' is being
+        compared against the 'dataset', which is only passed in the 'reduce' call.
+        """
+        dataset, solver, _ = reduce_problem
+        modified_solver = eqx.tree_at(
+            lambda x: x.coreset_size, solver, len(dataset) + 1
+        )
+        with pytest.raises(ValueError, match=re.escape(self.OVERSIZE_MSG)):
+            modified_solver.reduce(dataset)
+
+
+class TestKernelHerding(RefinementSolverTest, ExplicitSizeSolverTest):
```

Comment:
Yes - I suppose the only thing we are not checking in this new instance is (directly) if the kernel similarity penalty is infinite or not (due to unique=True). 

It might be worth putting some comments in before `expected_coreset` to explain where this analytical solution comes from - a much cut down version of the comments already in the `test_greedy_body` test?